### PR TITLE
Fix Incorrect turn lanes #18610

### PR DESCRIPTION
**Pull request by Dima-1**

### CODE REVIEW

File: OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java [hunk 1]
```diff
@@ -1766,11 +1766,15 @@ protected TurnType createSimpleKeepLeftRightTurn(boolean leftSide, RouteSegmentR
 					if (laneType != TurnType.C) {
 						lanes[it] = TurnType.getNext(laneType) << 1;
 					} else {
-						lanes[it] = TurnType.C << 1;
+						if (it == rs.leftLanes + currentLanesCount && currentLanesCount < prevLanesCount) {
+							lanes[it] = getTurnByAngle(rs.attachedAngles.get(rs.leftLanes + it - currentLanesCount)) << 1;
```
Comment (ivanPyrohivskyi) [inline OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java:1770]: rs.leftLanes + it - currentLanesCount == 2 * rs.leftLanes
is it correct ?

File: OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java [hunk 2]
```diff
@@ -1766,11 +1766,15 @@ protected TurnType createSimpleKeepLeftRightTurn(boolean leftSide, RouteSegmentR
...
+						} else {
+							lanes[it] = TurnType.C << 1;
+						}
 					}
 				} else {
 					// active lane
-					if (currentLanesCount == 1) {
+					if (currentLanesCount == 1 && rs.rightLanes == 0 ) {
```
Comment (ivanPyrohivskyi) [inline OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java:1777]: currentLanesCount == 1 && prevLinesCount == 1

File: OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java [hunk 3]
```diff
@@ -1766,11 +1766,15 @@ protected TurnType createSimpleKeepLeftRightTurn(boolean leftSide, RouteSegmentR
 					if (laneType != TurnType.C) {
 						lanes[it] = TurnType.getNext(laneType) << 1;
 					} else {
-						lanes[it] = TurnType.C << 1;
+						if (it == rs.leftLanes + currentLanesCount && currentLanesCount < prevLanesCount) {
```
Comment (ivanPyrohivskyi) [inline OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java:1769]: prevLanesCount - currentLanesCount > 0

File: OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java [hunk 4]
```diff
@@ -1749,35 +1751,42 @@ protected TurnType createSimpleKeepLeftRightTurn(boolean leftSide, RouteSegmentR
...
+			for (int it = 0; it < rs.leftLanes; it++) {
+				if (laneType != TurnType.C) {
+					// Avoid repeat of turns, e.g., for TSLL, get TL
+					lanes[it] = TurnType.getPrev(laneType) << 1;
+				} else {
+					if (currentLanesCount + rs.rightLanes < prevLanesCount) {
+						lanes[it] = getTurnByAngle(rs.attachedAngles.get(it)) << 1;
 					} else {
+						// Can be several straight directions
 						lanes[it] = TurnType.C << 1;
```
Comment (ivanPyrohivskyi) [inline OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java:1767]: Here must by secondary turn (glued to active)

File: OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java [hunk 5]
```diff
@@ -1749,35 +1751,42 @@ protected TurnType createSimpleKeepLeftRightTurn(boolean leftSide, RouteSegmentR
...
+			lanes = new int[totalLanes];
+
+			// Loop for lanes left from active
+			for (int it = 0; it < rs.leftLanes; it++) {
+				if (laneType != TurnType.C) {
+					// Avoid repeat of turns, e.g., for TSLL, get TL
+					lanes[it] = TurnType.getPrev(laneType) << 1;
+				} else {
+					if (currentLanesCount + rs.rightLanes < prevLanesCount) {
+						lanes[it] = getTurnByAngle(rs.attachedAngles.get(it)) << 1;
```
Comment (ivanPyrohivskyi) [inline OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java:1764]: getTurnByAngle must use for both cases

File: OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java [hunk 6]
```diff
@@ -1797,9 +1806,14 @@ public int compare(Double c1, Double c2) {
 		boolean allStraight = rs.allAreStraight();
 		int[] lanes = new int[1];
 		int extraLanes = 0;
+		double prevAngle = Double.NaN;
 		// iterate from left to right turns
 		for (int i = size - 1; i >= 0; i--) {
 			double angle = rs.attachedAngles.get(i);
+			if (angle == prevAngle) {
```
Comment (ivanPyrohivskyi) [inline OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java:1813]: Incorrect compare Double.NaN != Double.NaN, must be check also

File: OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java [hunk 7]
```diff
@@ -1749,35 +1751,42 @@ protected TurnType createSimpleKeepLeftRightTurn(boolean leftSide, RouteSegmentR
...
+			int totalLanes = rs.leftLanes + currentLanesCount + rs.rightLanes;
+			lanes = new int[totalLanes];
+
+			// Loop for lanes left from active
+			for (int it = 0; it < rs.leftLanes; it++) {
+				if (laneType != TurnType.C) {
+					// Avoid repeat of turns, e.g., for TSLL, get TL
+					lanes[it] = TurnType.getPrev(laneType) << 1;
+				} else {
+					if (currentLanesCount + rs.rightLanes < prevLanesCount) {
```
Comment (ivanPyrohivskyi) [inline OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java:1763]: Here must be <=

**Comment by ivanPyrohivskyi** [comment]
Cases for count left inactive lanes:
First number - currentLanesCount (ignore rights), second number - prevLanesCount
Count of leftLanes are 1 or 2
Малюнок без назви

**Comment by vshcherb** [comment]
Implemented slightly different approach with merged lanes

**Pull request closed without merging**